File: go/streaming/internal/stream_asr_job_incoming_message/audio_fragment_submission_progress_message.go
package stream_asr_job_incoming_message

type AudioFragmentSubmissionProgressMessage struct {
	Message string                                     `json:"message" validate:"eq=audio_fragment_submission_progress"`
	Body    AudioFragmentSubmissionProgressMessageBody `json:"body" validate:"required"`
}

type AudioFragmentSubmissionProgressMessageBody struct {
	Channels []AudioFragmentSubmissionProgressMessageBodyChannel `json:"channels"`
}

type AudioFragmentSubmissionProgressMessageBodyChannel struct {
	ChannelIndex       int `json:"channel_index" validate:"required"`
	AudioFragmentCount int `json:"audio_fragment_count" validate:"required"`
}

File: cpp/src/poetics_sdk/streaming/asr_job/incoming_message/audio_fragment_submission_progress_message.cpp
#include "audio_fragment_submission_progress_message.h"

namespace poetics::streaming::asr_job::incoming_message {
    AudioFragmentSubmissionProgressMessageBodyChannel::AudioFragmentSubmissionProgressMessageBodyChannel() : channelIndex(0), audioFragmentCount(0) {}
    AudioFragmentSubmissionProgressMessageBodyChannel::AudioFragmentSubmissionProgressMessageBodyChannel
        (int channelIndex, int audioFragmentCount) : channelIndex(channelIndex), audioFragmentCount(audioFragmentCount) {}

    AudioFragmentSubmissionProgressMessageBody::AudioFragmentSubmissionProgressMessageBody() {}
    AudioFragmentSubmissionProgressMessageBody::AudioFragmentSubmissionProgressMessageBody
        (std::vector<AudioFragmentSubmissionProgressMessageBodyChannel> channels) : channels(channels) {}

    AudioFragmentSubmissionProgressMessage::AudioFragmentSubmissionProgressMessage() : message(""), body() {}
    AudioFragmentSubmissionProgressMessage::AudioFragmentSubmissionProgressMessage
        (string message, AudioFragmentSubmissionProgressMessageBody body) : message(message), body(body) {}

    
    void from_json(const json& j, AudioFragmentSubmissionProgressMessageBody& body) {
        for (auto& channel : j["channels"]) {
            AudioFragmentSubmissionProgressMessageBodyChannel channelBody
                (channel["channel_index"], channel["audio_fragment_count"]);
            body.channels.push_back(channelBody);
        }
    }

    void from_json(const json& j, AudioFragmentSubmissionProgressMessage& message) {
        message.message = j["message"].get<string>();
        from_json(j["body"], message.body);
    }

    void from_json(const json& j, AudioFragmentSubmissionProgressMessageBodyChannel& channel) {
        channel.channelIndex = j["channel_index"].get<int>();
        channel.audioFragmentCount = j["audio_fragment_count"].get<int>();
    }
}
